src/test/java/io/vertx/core/net/NetTest.java

```diff
@@ -3368,6 +3394,104 @@ public void testHalfCloseCallsEndHandlerAfterBuffersAreDelivered() throws Except
+    server.close();
+    client.close();
+
+    // set up a normal server to force the SSL handshake time out in client
+    NetServerOptions serverOptions = new NetServerOptions()
+      .setSsl(false);
+    server = vertx.createNetServer(serverOptions);
+
+    NetClientOptions clientOptions = new NetClientOptions()
+      .setSsl(true)
+      .setTrustAll(true)
+      .setSslHandshakeTimeout(200)
+      .setSslHandshakeTimeoutUnit(TimeUnit.MILLISECONDS);
+    client = vertx.createNetClient(clientOptions);
+
+    server.connectHandler(s -> {
+    }).listen(testAddress, ar -> {
+      assertTrue(ar.succeeded());
+      client.connect(testAddress, res -> {
+        assertTrue(res.failed());
+        assertTrue(res.cause() instanceof SSLHandshakeException);
+        assertEquals("handshake timed out", res.cause().getCause().getMessage());
+        testComplete();
+      });
+    });
+    await();
+  }
+
+  @Test
+  public void testSslHandshakeTimeoutNotHappened() throws Exception {
+    server.close();
+    client.close();
+
+    NetServerOptions serverOptions = new NetServerOptions()
+      .setSsl(true)
+      .setKeyStoreOptions(Cert.SERVER_JKS.get())
+      // set 100ms to let the connection established
+      .setSslHandshakeTimeout(100)
+      .setSslHandshakeTimeoutUnit(TimeUnit.MILLISECONDS);
+    server = vertx.createNetServer(serverOptions);
+
+    NetClientOptions clientOptions = new NetClientOptions()
+      .setSsl(true)
+      .setTrustAll(true);
+    client = vertx.createNetClient(clientOptions);
+
+    server.connectHandler(s -> {
+    }).listen(testAddress, ar -> {
+      assertTrue(ar.succeeded());
+      client.connect(testAddress, res -> {
+        assertTrue(res.succeeded());
+        testComplete();
+      });
+    });
+    await();
+  }
+
+  @Test
+  public void testSslHandshakeTimeoutHappenedWhenUpgradeSsl() {
+    server.close();
+    client.close();
+
+    // set up a normal server to force the SSL handshake time out in client
+    NetServerOptions serverOptions = new NetServerOptions()
+      .setSsl(false);
+    server = vertx.createNetServer(serverOptions);
+
+    NetClientOptions clientOptions = new NetClientOptions()
+      .setSsl(false)
+      .setTrustAll(true)
+      .setSslHandshakeTimeout(200)
+      .setSslHandshakeTimeoutUnit(TimeUnit.MILLISECONDS);
+    client = vertx.createNetClient(clientOptions);
+
+    server.connectHandler(s -> {
+    }).listen(testAddress, ar -> {
+      assertTrue(ar.succeeded());
+      client.connect(testAddress, res -> {
+        assertTrue(res.succeeded());
+        NetSocket socket = res.result();
+
+        assertFalse(socket.isSsl());
+        socket.upgradeToSsl(v -> {
+          // this handler will never be called because of failure of handshake
+          fail("this should never be called");
+          assertFalse(socket.isSsl());
+        });
+        // wait a bit to make sure the above handler will not be called
+        vertx.setTimer(2000, id -> {
```

Comment:
this test should not use a timer, instead it should set an exception handler on the NetSocket which will get called when the upgrade fails, this is faster and more reliable.

Reply:
have a look at NetTest#TLSTest